### PR TITLE
[PyTorch][Distributed] Integrate with ShardedOptimizer in the unit test of ShardedLinear

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69569
* #69725

Since ShardedOptimizer is added in https://github.com/pytorch/pytorch/pull/68607. We now integrate it in our unit test for Sharded Linear.

Differential Revision: [D32777020](https://our.internmc.facebook.com/intern/diff/D32777020/)

### CODE REVIEW

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 1]
```diff
@@ -84,14 +88,28 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
...
+        self.assertNotEqual(previous, local_linear.weight)
+        previous_sharded = sharded_weight.clone().detach()
+        sharded_optim = ShardedOptimizer(dict(named_params_with_sharded_tensor(sharded_linear)), torch.optim.SGD, lr=0.1)
+        sharded_optim.step()
+        sharded_weight = sharded_linear.weight.local_shards()[0].tensor
+        local_weight_narrowed = local_linear.weight.narrow(
+            sharded_dim, start_pos, chunk_size
+        )
+        self.assertEqual(sharded_weight.size(), local_weight_narrowed.size())
+        with torch.no_grad():
```
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/ops/test_linear.py:109]: Why do we need `torch.no_grad` here?

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 2]
```diff
@@ -84,14 +88,28 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
...
+            self.assertNotEqual(previous_sharded, sharded_weight)
+            self.assertEqual(sharded_weight, local_weight_narrowed)
```
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/ops/test_linear.py:111]: Shouldn't we verify the bias as well?

File: test/distributed/_sharded_tensor/ops/test_linear.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:109]: It complains if not added. Because here `previous_sharded` is from detach() which does not come with gradient. And looks like the result from narrow also does not have gradient.

File: test/distributed/_sharded_tensor/ops/test_linear.py (hunk 2 above)
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:111]: sure.

**Comment by fduwjj** [comment]
Addressed the feedback from reviewers. This PR is dependent on https://github.com/pytorch/pytorch/pull/69725 and https://github.com/pytorch/pytorch/pull/68607. The test might fail for now but I patch dependent and test locally, the test passes.

**Review by wanchaol — APPROVED** [review]
lgtm, thanks!
